EDU-18914: Add Content modeling and architecture guide for headless stores (#2807)
vtexdocs/dev-portal-content: docs/guides/CMS/Integrations/Headless/content-modeling-and-architecture-for-headless-stores.md

---
title: "Content modeling and architecture for headless stores"
hidden: false
slug: "content-modeling-and-architecture-for-headless-stores"
createdAt: "2026-07-01T12:00:00.000Z"
updatedAt: "2026-07-01T12:00:00.000Z"
excerpt: "Learn how to model and consume CMS content in headless storefronts: JSON Schema structure, Content Types, components, and Data Plane API consumption patterns."
---

> Content modeling defines what can be created on the CMS. Architecture defines how that content moves from your repository to the CMS Admin and, after publishing, to your storefront.

Author: ```suggestion
Content modeling defines what can be created on the CMS and the architecture defines how that content moves from your repository to the CMS Admin and, after publishing, to your storefront.
```

> This guide covers modeling for headless integrations: core concepts, design principles, JSON Schema rules, consumption patterns, and recommended Content Type structures.

Author: ```suggestion
 This guide details both layers for headless integrations, covering core concepts, design principles, JSON Schema rules, consumption patterns, and recommended Content Type structures.
```

}
```

Learn more in the [JSON Schema step-by-step guide](https://json-schema.org/learn/getting-started-step-by-step).

### Content plugin (CLI)

Schemas are authored as `.jsonc` files in your project and uploaded through the [Content plugin](https://developers.vtex.com/docs/guides/content-plugin) (`@vtex/cli-plugin-content`):

```shell
vtex plugins install @vtex/cli-plugin-content
```

Typical workflow:

1. Author component and Content Type files in your repository.  
2. Run `vtex content generate-schema` to produce a unified bundle.  
3. Run `vtex content upload-schema` to publish it to the Schema Registry.

For headless stores, the `generate-schema` command requires two extra arguments: the paths to your component and Content Type directories, and `--base vtex.headless`:

```shell
vtex content generate-schema <components-dir> <pages-dir> --out <output-file> --base vtex.headless
```

For example:

```shell
vtex content generate-schema cms/mystore/components cms/mystore/pages --out cms/mystore/schema.json --base vtex.headless
```

### Headless store setup

| Requirement | Details |
| :---- | :---- |
| CMS enabled | CMS Admin available in your VTEX account. |

> | Headless store | A store created with storefront type Headless in **Storefront \> Content**. |

Author: ```suggestion
| Headless store | A store created with storefront type Headless in **Storefront > Content**. |
```

  CT["Content Type"]
  Comp["Component"]
  Sections["sections array"]
  Fields["Fields"]

  Schema -->|"defines"| CT
  Schema -->|"defines"| Comp
  CT --> Sections
  Sections -->|"references"| Comp
  Comp --> Fields
```

**Relationship in practice:**

| Layer | Declares | Example |
| :---- | :---- | :---- |
| Content Type | Page structure and which components are available | `landingPage` with `slug`, `seo`, and `sections` |
| Component | Reusable block with its own fields | `PromoBanner` with `title`, `image`, `link` |
| Field | Atomic editable value | `title` (string), `image.src` (string \+ media widget) |

A Content Type holds fixed fields (always present on every entry) and a dynamic `sections` array (you choose which components appear and in what order).

### Using a Content Type or a component

| Scenario | Content Type | Component |
| :---- | :---- | :---- |
| Represents a routable **page** | ✅ | |
| Needs **multiple instances** (many landing pages) | ✅ | |
| Only **one instance** store-wide | ✅ (`$singleton: true`) | |
| **Reusable UI block** on one or more pages | | ✅ |
| **Nested inside** another component | | ✅ |
| Needs a **slug** or identifier field | ✅ | |

> | Shared across the store (header, footer) | ✅ (singleton Content Type) | ✅ (component inside it) |

Reviewer: Colocar o X nesses espaços vazios
